add fcn to convert stdout from mc_ls fcn to a data frame

mskyttner opened the pull request:
The idea here is to allow for the directory listing output to be treated as a data frame, for example:

```
> mc_ls_tbl("play/")
# A tibble: 88 × 5
   timestamp           bytes size  attribs filename                             
   <dttm>              <dbl> <chr> <chr>   <chr>                                
 1 2023-07-17 03:50:30     0 0 B   NA      00-zhangjiakou-oss/                  
 2 2023-07-20 08:17:19     0 0 B   NA      2063b651-92a3-4a20-a4a5-03a96e7c5a89/
 3 2023-07-20 09:54:07     0 0 B   NA      64b8e8190e08712457ca5d28-files/      
 4 2023-07-20 13:03:13     0 0 B   NA      ameychoudharyresearchiiitacin/       
 5 2023-07-20 08:49:03     0 0 B   NA      ansar/                               
 6 2023-07-20 10:57:05     0 0 B   NA      asd/                                 
 7 2023-07-20 11:05:03     0 0 B   NA      asiatrip/                            
 8 2023-07-20 08:20:23     0 0 B   NA      backup/                              
 9 2023-07-20 08:45:25     0 0 B   NA      bsc-archive/                         
10 2023-07-20 12:04:34     0 0 B   NA      bucket/                              
# ℹ 78 more rows
# ℹ Use `print(n = ...)` to see more rows
```

To get the latest file in a bucket or find the biggest one:

```r
library(dplyr)
"play/email" |> mc_ls_tbl(recursive = T) |> arrange(desc(timestamp))
"play/email" |> mc_ls_tbl(recursive = T) |> arrange(desc(bytes))
```

cboettig commented:
@mskyttner thanks, this is a great idea.  

Minor fix I'll note inline as well -- but `mc_ls()` helper ought to have kept the `verbose()` argument to pass to `mc`, then you can set `verbose=FALSE` to suppress output and avoid needing to resort to `suppressMessages()` I think.  


I'm wondering if it is worth having a more generic parser for the `stdout` content that might be useful with other functions?  (I agree that `ls` is the most obvious method where you want to make use of the return contents downstream).

cboettig commented on R/mc_cat.R:
```diff
@@ -0,0 +1,39 @@
+  
+  cmd <- paste("cat", flags, target)
+  cmd <- gsub("\\s+", " ", cmd)
+  res <- suppressMessages(mc(cmd))
```
I think `res <- mc(cmd, verbose=FALSE)` would be cleaner than `suppressMessages()` here.

mskyttner replied:
Yes, makes sense, I tried to change it.

cboettig commented on R/mc_ls.R:
```diff
@@ -25,3 +25,85 @@ mc_ls <- function(target, recursive = FALSE, flags = "") {
+  
+  out <- readLines(con)
+  parse_mc_ls(trimws(out[nchar(out) > 0]))
+  
```
Sorry if I'm missing it, but I the logic of the connection parsing seems a bit overkill here?  Modulo the empty last line, I believe that this is identical to the simpler approach of doing `strplit()` on the `stdout` text string?

Compare: 

```r
res <- mc("ls play/", verbose=FALSE)
con <- textConnection(encoding = "UTF-8", object = suppressMessages(
  res$stdout
))
con
out <- readLines(con)
out <- out[vapply(out, nchar, 1) > 0]

out2 <- strsplit(res$stdout, "\n")[[1]]

identical(out, out2)
```

also, note that `nchar(out)` isn't vectorized, so might not be doing what you want it to do here?

mskyttner replied:
Yes, thanks! I tried simplify accordingly.

cboettig commented on R/mc_head.R:
```diff
@@ -0,0 +1,33 @@
+  
+  cmd <- paste("head", flags, target)
+  cmd <- gsub("\\s+", " ", cmd)
+  res <- suppressMessages(mc(cmd))
```
same as above, I think this wants to be `mc(cmd, verbose=FALSE)`

cboettig commented on R/mc_cat.R:
```diff
@@ -0,0 +1,39 @@
+  on.exit(close(con))
+  
+  out <- readLines(con)
+  paste(collapse = "\n", out)
```
using readLines and handling a connection object seems overkill, I think the result of line 38 is identical to `res$stdout` text we began with?

mskyttner replied:
Yeah, looks I was doing some unnecessary "round-tripping" there... I tried to clean it up below.

cboettig reviewed (changes requested):
This is really nice, very useful and I know writing string parsing code sucks, so thanks for doing this.  Lemme know if I'm missing something about the readLines() bit, I think it can be simplified as noted inline?

mskyttner commented:
Thanks for considering and for the feedback provided! 

I discovered the mc commands have a global --json flag which returns some more information for directory listings (but/and is ignored for "mc cat" and "mc head" invocations). That format is probably more standardized across clients/OS-platforms (I'm not sure how various locale settings come into play when using the minio client "standard" stdout, for example with regards to timezones in timestamps). 

I tried adding a parameter to use the json-lines format and also fixed some issues I saw with various variants of listings (local vs remote, recursive or not) and added a few more tests.

mskyttner commented:
Maybe the default should be to use the jsonl format...? I think that the timestamps are in UTC there and in my local timezone otherwise:

```r
> mc_ls_tbl(".", use_json = TRUE) |> filter(is_folder) |> arrange(desc(last_modified)) 
# A tibble: 7 × 9
  key          last_modified       bytes size  storage_class is_folder   ver etag  url                            
  <chr>        <dttm>              <int> <chr> <chr>         <lgl>     <int> <chr> <chr>                          
1 .git/        2023-07-25 10:31:17  4096 4 KiB NA            TRUE          1 NA    /home/markus/repos/minioclient/
2 R/           2023-07-20 14:47:38  4096 4 KiB NA            TRUE          1 NA    /home/markus/repos/minioclient/
3 man/         2023-07-20 14:47:38  4096 4 KiB NA            TRUE          1 NA    /home/markus/repos/minioclient/
4 .Rproj.user/ 2023-07-20 07:43:23  4096 4 KiB NA            TRUE          1 NA    /home/markus/repos/minioclient/
5 .github/     2023-07-20 07:43:08  4096 4 KiB NA            TRUE          1 NA    /home/markus/repos/minioclient/
6 inst/        2023-07-20 07:43:08  4096 4 KiB NA            TRUE          1 NA    /home/markus/repos/minioclient/
7 tests/       2023-07-20 07:43:08  4096 4 KiB NA            TRUE          1 NA    /home/markus/repos/minioclient/

> mc_ls_tbl(".") |> filter(is_folder) |> arrange(desc(last_modified)) 
# A tibble: 7 × 6
  key          last_modified       bytes size  storage_class is_folder
  <chr>        <dttm>              <dbl> <chr> <chr>         <lgl>    
1 .git/        2023-07-25 12:31:17  4096 4 KiB NA            TRUE     
2 R/           2023-07-20 16:47:38  4096 4 KiB NA            TRUE     
3 man/         2023-07-20 16:47:38  4096 4 KiB NA            TRUE     
4 .Rproj.user/ 2023-07-20 09:43:23  4096 4 KiB NA            TRUE     
5 .github/     2023-07-20 09:43:08  4096 4 KiB NA            TRUE     
6 inst/        2023-07-20 09:43:08  4096 4 KiB NA            TRUE     
7 tests/       2023-07-20 09:43:08  4096 4 KiB NA            TRUE     
```

cboettig commented on R/mc_ls.R:
```diff
@@ -17,11 +19,166 @@
+        target = target, 
+        recursive = recursive, 
+        verbose = FALSE, 
+        flags = "--json"
```
json flag, brilliant! :tada:

mskyttner commented:
A few thoughts on the function signature:

Regarding specifically mc_ls(), I think it is similar to base::dir() in that offering the recursive parameter feels natural to either traverse or not on the server side. Thinking of the fs package, this param is called recurse in fs::dir_ls() (previously "recursive" there, it seems?), and this function also offers filtering with regexp and ability to invert results. But such file searching features would probably be more natural to offer via a "mc_find()" function(?) (via the "mc find" command) so they execute on the server side rather than on the client side (where it can be done in several different ways anyhow).

Since the --json output seems to offer a pareto improvement over parsing the stdout, as you noted, I would suggest removing the use_json parameter altogether and use (only) it by default. So I would suggest to remove the variant which does more string parsing on the non-jsonl stdout return value, so it doesn't need to be maintained. It might be nice to have a way to get directly the raw json, but that is provided via the "mc"-function anyways if someone wants to do their own parsing of the jsonlines.
    
Personally, I tend to want my directory listings with more than just the filename and would prefer to have a data frame as the default return, something like what you get from "ls -lahtr" in a shell. Some other packages, like in fs::dir_ls() and in base R base::dir(), you get just the filenames when requesting a directory listing. For a richer data frame return, like, one could compare to base::file.info() or the more modern fs::file_info() where size is given fs::as_fs_bytes:
  
```r
> file.info(".", extra_cols = T) |> str()
'data.frame':	1 obs. of  10 variables:
 $ size  : num 4096
 $ isdir : logi TRUE
 $ mode  : 'octmode' int 775
 $ mtime : POSIXct, format: "2023-07-26 08:19:45"
 $ ctime : POSIXct, format: "2023-07-26 08:19:45"
 $ atime : POSIXct, format: "2023-07-26 08:19:46"
 $ uid   : int 1000
 $ gid   : int 1000
 $ uname : chr "markus"
 $ grname: chr "markus"

> fs::file_info(".") |> str()
tibble [1 × 18] (S3: tbl_df/tbl/data.frame)
 $ path             : 'fs_path' chr "."
 $ type             : Factor w/ 8 levels "any","block_device",..: 4
 $ size             : 'fs_bytes' num 4K
 $ permissions      : 'fs_perms' int rwxrwxr-x
 $ modification_time: POSIXct[1:1], format: "2023-07-26 08:19:45"
 $ user             : chr "markus"
 $ group            : chr "markus"
 $ device_id        : num 64769
 $ hard_links       : num 9
 $ special_device_id: num 0
 $ inode            : num 53477819
 $ block_size       : num 4096
 $ blocks           : num 8
 $ flags            : int 0
 $ generation       : num 0
 $ access_time      : POSIXct[1:1], format: "2023-07-26 08:19:46"
 $ change_time      : POSIXct[1:1], format: "2023-07-26 08:19:45"
 $ birth_time       : POSIXct[1:1], format: "2023-07-20 09:43:06"
```

Regarding the parameter currently named show_fullpath, I think it is nice to be able to get the absolute path for a file/object, probably by default. For constructing a full path, the --json output provides a "url" field, but it does not point directly to the object... I think really is more like a parent node or base_url... the key needs to be appended to get the full path to the object/resource. Since it is not really always a url (or maybe it is, I was thinking of when listing local files but an absolute local file path may still count as a url, at least when used with functions such as base::browseURL()) I think "dirpath" (or "dir" or "basepath") could be a more intuitive name for it that works both for local and remote listings.

Maybe the show_fullpath should be always on, ie not optional, so the parameter should be removed?... but it depends how much convenience it gives versus how unnecessary it is considered - given that key and dirpath is avaiable and can be combined into the fullpath easily, what do you think?

With the latest changes, I suggest to additionally by default to provide the fullpath which immediately provides the absolute path to the object, locally or remotely:

  
```r    
> mc_mirror("play/test-bucket", "/tmp/test-bucket")
  
> mc_ls("play/test-bucket", recursive = T) |> select(fullpath, dirpath)
# A tibble: 6 × 2
  fullpath                                        dirpath                         
  <chr>                                           <chr>                           
1 https://play.min.io/test-bucket/object-1        https://play.min.io/test-bucket/
2 https://play.min.io/test-bucket/object-2        https://play.min.io/test-bucket/
3 https://play.min.io/test-bucket/object-3        https://play.min.io/test-bucket/
4 https://play.min.io/test-bucket/output (2).docx https://play.min.io/test-bucket/
5 https://play.min.io/test-bucket/owl (2) (1).png https://play.min.io/test-bucket/
6 https://play.min.io/test-bucket/test-object     https://play.min.io/test-bucket/

> mc_ls("/tmp/test-bucket", recursive = T) |> select(fullpath, dirpath)
# A tibble: 6 × 2
  fullpath                         dirpath          
  <chr>                            <chr>            
1 /tmp/test-bucket/object-1        /tmp/test-bucket/
2 /tmp/test-bucket/object-2        /tmp/test-bucket/
3 /tmp/test-bucket/object-3        /tmp/test-bucket/
4 /tmp/test-bucket/output (2).docx /tmp/test-bucket/
5 /tmp/test-bucket/owl (2) (1).png /tmp/test-bucket/
6 /tmp/test-bucket/test-object     /tmp/test-bucket/
```

Since it seems several targets can be specified to the "mc ls" command, this works (providing a vector of target "urls"):
    
```r
>   mc_ls(c("/tmp/test-bucket", "play/test-bucket")) |> select(key, etag, fullpath)
# A tibble: 12 × 3
   key             etag                             fullpath                                       
   <chr>           <chr>                            <chr>                                          
 1 object-1        NA                               /tmp/test-bucket/object-1                      
 2 object-2        NA                               /tmp/test-bucket/object-2                      
 3 object-3        NA                               /tmp/test-bucket/object-3                      
 4 output (2).docx NA                               /tmp/test-bucket/output (2).docx               
 5 owl (2) (1).png NA                               /tmp/test-bucket/owl (2) (1).png               
 6 test-object     NA                               /tmp/test-bucket/test-object                   
 7 object-1        3d4c0a212908ea883221a0127466f42c https://play.min.io/test-bucket/object-1       
 8 object-2        3d4c0a212908ea883221a0127466f42c https://play.min.io/test-bucket/object-2       
 9 object-3        3d4c0a212908ea883221a0127466f42c https://play.min.io/test-bucket/object-3       
10 output (2).docx d6fda183c13b71ff080a1877031842f6 https://play.min.io/test-bucket/output (2).docx
11 owl (2) (1).png 1cc789c3b5241cc6b80b4c414d098ed9 https://play.min.io/test-bucket/owl (2) (1).png
12 test-object     3d4c0a212908ea883221a0127466f42c https://play.min.io/test-bucket/test-object   
```

cboettig commented:
Thanks for these updates and the perspectives here, really good points and much appreciated. (using the checkbox format to distinguish some comments of "i agree" from ones where maybe there's some follow up?)

- [x] Re `recurse` vs `recursive`, yeah, never quite understood why `fs` made that change (maybe a preference for verbs over adjectives to describe actions), but I think it makes most sense to retain the term used by `mc` (and everywhere else).  I do think recursion is a commonly used feature independent of other tasks where `find` is the better choice.

- [x] dropping the string parsing and just relying on `--json` makes sense, simpler codebase, easier to maintain.

- [x] it's nice you can use vector args for `mc_ls()` !  Thanks for updating the doc-string for that. 

- [ ] Very minor, re naming `url` vs `dirpath`, in `mc find`, this part of the full path can be referenced using the syntax `{dir}`, which I kinda like -- it's shorter than `dirpath` but not specific to urls.    

- [ ] Re vector vs table return (i.e. `fs::dir_ls()` vs `fs::dir_info()`)  --   one use case where the vector return is nice is in piped workflows, e.g. `fs::dir_ls(glob = something) |> fs::file_delete()` etc.   Maybe we ought to have an option to return just the key vs the table syntax? (kinda like `ls` vs `ls -l` in bash).   

I'm actually tempted to say that the vector format should be the default return type. The table takes up a lot of real estate on the screen (moreso with the default data.frame print method, though with the default tibble print method most columns are truncated).  For quick interactive use, it's nice that something concise like `mc_ls("play")` with no arguments gives a concise reply of just the keys -- no file paths, no table, while `mc_ls("play", details=TRUE)` or something like that shows the table?  (of course a user who wants to just see the keys can add  `|> dplyr::pull(key)`, but it feels a bit backward to me that the simpler behavior requires the more complex invocation).  (e.g. I'll often do a quick `mc_ls()` just to see what a key is in a bucket because I can't quite remember the key name that I need in some other call, or just want to "peek" at the keys in a bucket)

In `mc` these kind of piped operations are easy to do with `mc find --exec` syntax, but feels less intuitive / "R native" way.    
(we should add `mc_find` (in another PR), it's a particularly powerful function in `mc` thanks to it's  `--exec` option, but the syntax with `{}` is a bit novel to users and I wasn't sure if we'd want to wrap it in some way.)

mskyttner commented:
I tried to make some changes to accommodate the latest feedback... some thoughts that I had:

If a vector of keys is returned by default from mc_ls(), and one would like send those results to something like say mc_stat() to inspect metadata, one would like to get the path including the "minio alias" back or one would need to prepend it to the keys before sending it on in the pipe. Currently:

```r
> mc_ls("play") |> head()
[1] "00-test/"                              "000-adi-bucket/"                       "00vytautas/"                          
[4] "1e1ab0c2-104b-4aee-aa91-235003862a1c/" "2063b651-92a3-4a20-a4a5-03a96e7c5a89/" "365b3a0c-164d-4843-8357-fc08307513f3/"
```
If a vector of several target specifications is used in the call, the json return only has the url available for distinguishing what original targets are associated with the keys returned. I tried to provide something to get back the "minio relative target path" in the detailed data frame in the path column:

```r
> mc_ls("play", details = TRUE) |> head()
# A tibble: 6 × 12
  key                        last_modified       bytes size  storage_class is_folder   ver etag  dir   reldir path  abspath
  <chr>                      <dttm>              <int> <chr> <chr>         <lgl>     <int> <chr> <chr> <chr>  <chr> <chr>  
1 00-test/                   2023-07-27 15:50:45     0 0 B   NA            TRUE          1 NA    http… play/  play… https:…
2 000-adi-bucket/            2023-07-27 08:59:14     0 0 B   NA            TRUE          1 NA    http… play/  play… https:…
3 00vytautas/                2023-07-27 16:47:31     0 0 B   NA            TRUE          1 NA    http… play/  play… https:…
4 1e1ab0c2-104b-4aee-aa91-2… 2023-07-26 23:54:36     0 0 B   NA            TRUE          1 NA    http… play/  play… https:…
5 2063b651-92a3-4a20-a4a5-0… 2023-07-27 02:51:08     0 0 B   NA            TRUE          1 NA    http… play/  play… https:…
6 365b3a0c-164d-4843-8357-f… 2023-07-27 00:38:51     0 0 B   NA            TRUE          1 NA    http… play/  play… https:…
```
A call like the one below should be easier to send on in a pipe than one with just the "key" values?

```r
> mc_ls("play", details = TRUE)$path |>  head()
[1] "play/00-test/"                              "play/000-adi-bucket/"                      
[3] "play/00vytautas/"                           "play/1e1ab0c2-104b-4aee-aa91-235003862a1c/"
[5] "play/2063b651-92a3-4a20-a4a5-03a96e7c5a89/" "play/365b3a0c-164d-4843-8357-fc08307513f3/"
```

cboettig commented:
nice, that's a good point!  I like this a lot.  It does feel more consistent with the input syntax and more pipe-friendly this way.  

Going to just test drive this a little but I think we're good to merge now -- hope we can get it posted once the CRAN team is back from holiday!
